ws-server/src/future.rs

```diff
@@ -130,3 +136,80 @@ where
 		this.selector.poll_unpin(cx)
 	}
 }
+
+#[derive(Debug)]
+struct MonitorInner {
+	shutdown_requested: AtomicBool,
+	waker: AtomicWaker,
+}
+
+/// Monitor for checking whether the server has been flagged to shut down.
+#[derive(Debug, Clone)]
+pub(crate) struct StopMonitor(Arc<MonitorInner>);
+
+impl Drop for StopMonitor {
+	fn drop(&mut self) {
+		if Arc::strong_count(&self.0) == 1 {
+			self.0.waker.wake();
+		}
+	}
+}
+
+impl StopMonitor {
+	pub(crate) fn new() -> Self {
+		StopMonitor(Arc::new(MonitorInner { shutdown_requested: AtomicBool::new(false), waker: AtomicWaker::new() }))
+	}
+
+	pub(crate) fn shutdown_requested(&self) -> bool {
+		self.0.shutdown_requested.load(Ordering::Relaxed)
+	}
+
+	pub(crate) fn handle(&self) -> StopHandle {
+		StopHandle(Arc::downgrade(&self.0))
+	}
+}
+
+/// Handle that is able to stop the running server.
+#[derive(Debug, Clone)]
+pub struct StopHandle(Weak<MonitorInner>);
+
+impl StopHandle {
+	/// Requests server to stop. Returns an error if server was already stopped.
+	///
+	/// Returns a future that can be awaited for when the server shuts down.
+	pub fn stop(self) -> Result<ShutdownWaiter, Error> {
+		if let Some(arc) = Weak::upgrade(&self.0) {
+			// We proceed only if the previous value of the flag was `false`
+			if arc.shutdown_requested.swap(true, Ordering::Relaxed) == false {
```

Comment:
I'm not sure about the memory ordering here,  can't the `if let Some(arc...` be ordered any given order by the compiler or are we relying on that the `Weak::upgrade` itself is doing `atomic stuff`?